deploy/buildtools/buildconfig.py

```diff
@@ -1,177 +1,151 @@
-""" This converts the build configuration files into something usable by the
-manager """
-
 from time import strftime, gmtime
-import configparser
 import pprint
+from importlib import import_module
 
-from runtools.runtime_config import RuntimeHWDB
 from awstools.awstools import *
+from buildtools.buildfarmhostdispatcher import BuildFarmHostDispatcher
+
+# typing imports
+from typing import Set, Type, Any, Optional, Dict, TYPE_CHECKING
+# needed to avoid type-hint circular dependencies
+# TODO: Solved in 3.7.+ by "from __future__ import annotations" (see https://stackoverflow.com/questions/33837918/type-hints-solve-circular-dependency)
+#       and normal "import <module> as ..." syntax (see https://www.reddit.com/r/Python/comments/cug90e/how_to_not_create_circular_dependencies_when/)
+if TYPE_CHECKING:
+    from buildtools.buildconfigfile import BuildConfigFile
+else:
+    BuildConfigFile = object
+
+def inheritors(klass: Type[Any]) -> Set[Type[Any]]:
```

Comment:
So the big question i have as a relative python novice is the scope over which __subclasses__ will resolve subclasses. Some light googling suggests this will only resolve subclasses defined in the current scope. Here's some interesting discussion on this subject: https://stackoverflow.com/a/28056657

I think, unless there's a more rigorous mechanism to do dependency injection, we probably want to be able to let users add their own subclasses by glomming stuff onto the python path. 

In any case, the scope of this function should be described in the docstring.

Reply:
I like @timsnyder-siv general approach of putting the SO link the code for more arcane stuff. This one is marginal, but it's probably still a good idea?

Reply:
I think this is worth including the link as a comment in the code.

Reply:
> glomming stuff onto the python path

Probably want to take a look at https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/ and use one of the patterns there for discovering stuff that implements something we want to pull in as a plugin.  That way we can have something that imports the discovered plugins at runtime so that the `__subclasses__()` will correctly identify those as implementations of our junk.

Reply:
Oh, and I think it's worth putting the link to the SO post as a comment in the code.  Helps tremendously to give context and I know it isn't required but I like to attribute back to those things.

Reply:
Looking in more detail at https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/, I would suggest that we do the first option of just importing all toplevel modules that start with 'firesim_' and that way people can do the least amount of work to create a 'firesim_whatever' directory that they glom into their PYTHONPATH (or actually package and install) and it will get loaded and can provide custom implementations of the abstract base classes.

Reply:
SGTM on switching to a better mechanism. I think this should be done in a followup PR for now though. I'll leave this comment open and keep track of this in an issue when the PR merges.

Reply:
Ok this should be resolved now by removing `getattr`.